src/argilla/client/feedback.py

```diff
@@ -0,0 +1,807 @@
+            raise TypeError(f"Expected `guidelines` to be a string, got {type(guidelines)} instead.")
+        self.__guidelines = guidelines
+
+        if not isinstance(fields, list):
+            raise TypeError(f"Expected `fields` to be a list, got {type(fields)} instead.")
+        any_required = False
+        for field in fields:
+            if not isinstance(field, FieldSchema):
+                raise TypeError(f"Expected `fields` to be a list of `FieldSchema`, got {type(field)} instead.")
+            if not any_required and field.required:
+                any_required = True
+        if not any_required:
+            raise ValueError("At least one `FieldSchema` in `fields` must be required (`required=True`).")
+        self.__fields = fields
+        self.__fields_schema = None
+
+        if not isinstance(questions, list):
+            raise TypeError(f"Expected `questions` to be a list, got {type(questions)} instead.")
+        any_required = False
+        for question in questions:
+            if not isinstance(question, QuestionSchema):
+                raise TypeError(f"Expected `questions` to be a list of `QuestionSchema`, got {type(question)} instead.")
+            if not any_required and question.required:
+                any_required = True
+        if not any_required:
+            raise ValueError("At least one `QuestionSchema` in `questions` must be required (`required=True`).")
+        self.__questions = questions
+
+        self.__records = []
+        self.__new_records = []
+
+    def __len__(self) -> int:
+        """Returns the number of records in the dataset."""
+        return len(self.records)
+
+    def __getitem__(self, key: Union[slice, int]) -> Union["FeedbackItemModel", List["FeedbackItemModel"]]:
+        """Returns the record(s) at the given index(es).
+
+        Args:
+            key: the index(es) of the record(s) to return. Can either be a single index or a slice.
+
+        Returns:
+            Either the record of the given index, or a list with the records at the given indexes.
+        """
+        if len(self.records) < 1:
+            raise RuntimeError(
+                "In order to get items from `rg.FeedbackDataset` you need to either add"
+                " them first with `add_records` or fetch them from Argilla or"
+                " HuggingFace with `fetch_records`."
+            )
+        if isinstance(key, int) and len(self.records) < key:
+            raise IndexError(f"This dataset contains {len(self)} records, so index {key} is out of range.")
+        return self.records[key]
+
+    def __del__(self) -> None:
+        """When the dataset object is deleted, delete all the records as well to avoid memory leaks."""
+        if hasattr(self, "__records") and self.__records is not None:
+            del self.__records
+        if hasattr(self, "__new_records") and self.__new_records is not None:
+            del self.__new_records
+
+    def __enter__(self) -> "FeedbackDataset":
```

Comment:
What the use case you're thinking about the context manager?

Reply:
I used it a couple times while testing everything, but probably there's no use-cases, so we can remove this